libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -628,13 +680,55 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	expectDeposit(1, height);
 }
 
+void CodeTransform::allocateReturnSlotsIfNeeded(Statement const& _statement)
+{
+	if (m_returnStackHeight.has_value())
+		return;
+	if (holds_alternative<FunctionDefinition>(_statement))
+		return;
+	if (
+		holds_alternative<ExpressionStatement>(_statement) ||
+		holds_alternative<Assignment>(_statement)
+	)
+	{
+		ReferencesCounter referencesCounter{ReferencesCounter::CountWhat::OnlyVariables};
+		referencesCounter.visit(_statement);
+		auto isReferenced = [&](YulString _returnVariableName) {
+			return referencesCounter.references().count(_returnVariableName);
+		};
+		if (ranges::none_of(m_delayedReturnVariables, isReferenced))
+			return;
+	}
+
+	for (YulString _returnVariableName: m_delayedReturnVariables)
+	{
+		auto* var = std::get_if<Scope::Variable>(m_scope->lookup(_returnVariableName));
+		yulAssert(var, "Return variable not in scope.");
+		m_context->variableStackHeights[var] = static_cast<size_t>(m_assembly.stackHeight());
+		// Preset stack slots for return variables to zero.
+		m_assembly.appendConstant(u256(0));
+	}
+
+	m_delayedReturnVariables.clear();
+	m_returnStackHeight = m_assembly.stackHeight();
+	m_context->functionExitPoints.push(
+		CodeTransformContext::JumpInfo{
+			m_assembly.newLabelId(),
+			m_assembly.stackHeight()
+		}
+	);
+	m_unusedStackSlots.clear();
```

Comment:
Why do they have to be cleared?

Reply:
They don't *have* to, but none of the slots in there will ever be reused again afterwards, so it makes sense to clear them.